// Copyright 2023 The Confidential Federated Compute Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package pipeline_transform;

service PipelineTransform {
  // Transforms zero or more inputs into zero or more outputs. This method can
  // be used to implement map (1 -> 1), filter (1 -> 0 or 1), combine (N ->
  // 1), "do" (1 -> N), or more general (M -> N) operations.
  // method_id: 1
  rpc Transform(TransformRequest) returns (TransformResponse);
}

// A request providing the inputs to be transformed.
message TransformRequest {
  repeated Input inputs = 1;

  message Input {
    bytes data = 1;
  }
}

// A response providing the results of the transformation.
message TransformResponse {
  repeated Output outputs = 1;

  message Output {
    bytes data = 1;
  }
}
